pkg/controller/cluster/controller.go

```diff
@@ -199,32 +220,36 @@ func lateInitializeCluster(p *v1alpha1.ClusterParameters, r *argocdv1alpha1.Clus
 	if p.Shard == nil {
 		p.Shard = r.Shard
 	}
+
+	if p.Server == nil {
+		p.Server = &r.Server
+	}
+
+	if p.Name == nil {
+		p.Name = &r.Name
+	}
+
 }
 
-func generateClusterObservation(r *argocdv1alpha1.Cluster) v1alpha1.ClusterObservation {
+func generateClusterInfo(r *argocdv1alpha1.Cluster) v1alpha1.ClusterInfo {
 	if r == nil {
-		return v1alpha1.ClusterObservation{}
-	}
-
-	o := v1alpha1.ClusterObservation{
-		ClusterInfo: v1alpha1.ClusterInfo{
-			ConnectionState: (*v1alpha1.ConnectionState)(&r.Info.ConnectionState),
-			ServerVersion:   &r.Info.ServerVersion,
-			CacheInfo: &v1alpha1.ClusterCacheInfo{
-				ResourcesCount:    &r.Info.CacheInfo.ResourcesCount,
-				APIsCount:         &r.Info.CacheInfo.APIsCount,
-				LastCacheSyncTime: r.Info.CacheInfo.LastCacheSyncTime,
-			},
-			ApplicationsCount: r.Info.ApplicationsCount,
+		return v1alpha1.ClusterInfo{}
+	}
+	i := v1alpha1.ClusterInfo{
+		ConnectionState: (*v1alpha1.ConnectionState)(&r.Info.ConnectionState),
+		ServerVersion:   &r.Info.ServerVersion,
+		CacheInfo: &v1alpha1.ClusterCacheInfo{
+			ResourcesCount:    &r.Info.CacheInfo.ResourcesCount,
+			APIsCount:         &r.Info.CacheInfo.APIsCount,
+			LastCacheSyncTime: r.Info.CacheInfo.LastCacheSyncTime,
 		},
+		ApplicationsCount: r.Info.ApplicationsCount,
 	}
-
-	return o
+	return i
```

Comment:
Is this rework necessary? I would prefer to have all status updates in a single function that is called from `Observe`.

Reply:
I've reverted to the original return of `ClusterObservation` and added the update to the Kubeconfig Secret in this function.